keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,201 @@
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`/ `Windows Containers`, the way to collect metrics from
+these containers is quite different from runC. For the third-party metric
+collectors, they have to maintain the knowledge about how to collect metrics
+from each kind of container runtimes. And more, kubelet codebase focus on
+core-metrics export (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for different container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
+it is easy for the collector from monitor pipeline (defined in [Kubernetes monitoring architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) to collect metrics from different kind of runtimes.
+
+### Goals
+
+- Enhance Container Runtime Interface(CRI) to provide more metrics about
+managed container
+
+### Non-Goals
+
+- Extend Core-Metrics (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)) for 
```

Comment:
What do you mean by extend? Will this KEP address possibility of replacing Summary API?

Reply:
Sorry for confusing. I removed this part because the core metrics is related to kubernetes scheduler. This KEP is focusing on how to reduce burden on kubelet by collecting metrics from ContainerStats directly. Core-Metrics is other topic. 😂